Simulation Code/Base/Base.ino

```diff
@@ -1,112 +1,138 @@
+            } else {
+                digitalWrite(WaterLevelD, LOW);
+            }
+            pumpPWM = 0;
+            analogWrite(pin, pumpPWM);
+        } else if (pumpStatus && (elapsedPumpTime <= pumpTimeOn)) {
+            if (waterLevelA == HIGH && elapsedTime >= stabilityTime) {
+                pumpPWM = 255;
+            } else if (waterLevelB == HIGH && elapsedTime >= stabilityTime) {
+                pumpPWM = 170;
+            } else if (waterLevelC == HIGH && elapsedTime >= stabilityTime) {
+                pumpPWM = 85;
+            } else if (waterLevelC == LOW && elapsedTime >= stabilityTime) {
+                pumpPWM = 0;
+            }
+            analogWrite(pin, pumpPWM);
+            previousTime = currentTime;
+        } else if (elapsedPumpTime >= pumpTimeOn) {
+            digitalWrite(pin, LOW);
+            pumpStatus = false;
+            previousPumpTime = currentTime;
+        } else if (elapsedPumpTime >= pumpTimeOff) {
+            analogWrite(pin, pumpPWM);
+            pumpStatus = true;
+            previousPumpTime = currentTime;
+        }
     }
-    analogWrite(WaterPump, PumpPWM); // Apply the new pump speed
-  }
-  else if (elapsedPumpTime >= pumpTimeOn) // If the pump has been on for more than the specified on time
-  {
-    digitalWrite(WaterPump, LOW); // Turn off the pump
-    pumpStatus = false; // Update the pump status
-    previousPumpTime = currentPumpTime; // Update the time of the last pump status check
-  }
-  else if (elapsedPumpTime >= pumpTimeOff)// If the pump has been off for more than the specified off time
-  {
-    analogWrite(WaterPump, PumpPWM); // Turn on the pump
-    pumpStatus = true; // Update the pump status
-    previousPumpTime = currentPumpTime; // Update the time of the last pump status check
-  }
+};
+
+// Instantiate structures for each component
+WaterLevelSensors waterLevelSensors = {WaterLevelA, WaterLevelB, WaterLevelC};
+Fans fans = {FanA, FanB};
+WaterPumpController waterPumpController = {WaterPump};
+
+void setup() {
+    Serial.begin(9600);
+    waterLevelSensors.begin();
+    fans.begin();
+    waterPumpController.begin();
 }
 
-void loop() 
-{
-  int waterLevelA, waterLevelB, waterLevelC; // Variables to store the current water level
-  unsigned long currentTime, elapsedTime, currentPumpTime, elapsedPumpTime; // Variables to store the current time and elapsed time
+void loop() {
+    int waterLevelA, waterLevelB, waterLevelC;
+    unsigned long currentTime = millis();
 
-  checkWaterLevel(waterLevelA, waterLevelB, waterLevelC); // Check the current water level
-  checkPumpStatus(currentTime, elapsedTime, currentPumpTime, elapsedPumpTime, 
-  waterLevelA, waterLevelB, waterLevelC); // Check and update the pump status
-  
+    waterLevelSensors.readLevels(waterLevelA, waterLevelB, waterLevelC);
+    waterPumpController.updatePumpStatus(currentTime, waterLevelA, waterLevelB, waterLevelC, WaterLevelD);
+    fans.update(currentTime);
```

Comment:
Lines 108-111 and 135-137 essentially do the same thing. What is the added benefit of creating a structure to redefine these functions?